python_modules/libraries/dagster-dbt/dagster_dbt/dbt_artifacts.py

```diff
@@ -0,0 +1,152 @@
+import logging
+import os
+import shutil
+from functools import cached_property
+from pathlib import Path
+from typing import List, Optional, Union
+
+from dagster._annotations import public
+
+from .core.resources_v2 import DbtCliResource
+from .errors import DagsterDbtManifestNotPreparedError, DagsterDbtProjectNotFoundError
+
+
+class DbtArtifacts:
+    def __init__(
+        self,
+        project_dir: Union[Path, str],
+        *,
+        target_folder: Union[Path, str] = "target",
+        prepare_command: List[str] = ["parse", "--quiet"],
+        package_data_dir: Optional[Union[Path, str]] = None,
+    ):
+        """A utility class to help manage dbt artifacts in different deployment contexts.
+
+        This class provides a setup to solve for these goals:
+        * During development, reload the manifest at run time to pick up any changes.
+        * When deploying, expect a manifest that was prepared at build time to reduce start-up time.
+        * Handle the scenario when the dbt project is copied in to a directory for packaging.
+
+        Args:
+            project_dir (Union[str, Path]):
+                The directory of the dbt project.
+            target_folder (Union[str, Path]):
+                The folder in the project project directory to output artifacts.
+                Default: "target"
+            prepare_command: The dbt cli command to run to prepare the manifest.json
+                Default: ["parse", "--quiet"]
+            package_data_dir (Union[str, Path]):
+                A directory that will contain a copy of the dbt project, synced at build time.
+        """
+        self.project_dir = Path(project_dir)
+        if not self.project_dir.exists():
+            raise DagsterDbtProjectNotFoundError(f"project_dir {project_dir} does not exist.")
+        self._target_folder = Path(target_folder)
+        self._prepare_command = prepare_command
+        # this is ok if it doesn't exist, will get created by `prepare`
+        self._package_data_dir = Path(package_data_dir) if package_data_dir else None
+
+    @public
+    @cached_property  # to prevent repeat compilation
+    def manifest_path(self) -> Path:
```

Comment:
I believe it's problematic for something that looks like a property (which should be inexpensive) to trigger something as expensive as a compile step. Why not check this at object construction time?